instrumentation/opentelemetry_req/lib/opentelemetry_req.ex

```diff
@@ -139,98 +291,162 @@ defmodule OpentelemetryReq do
   defp build_req_attrs(request) do
     uri = request.url
     url = sanitize_url(uri)
+    config = Req.Request.get_private(request, :otel)
+
+    opt_in = %{
+      HTTPAttributes.http_request_body_size() => extract_request_body_size(request),
+      NetworkAttributes.network_transport() => :tcp,
+      URLAttributes.url_scheme() => extract_scheme(uri),
+      URLAttributes.url_template() => extract_url_template(request),
+      UserAgentAttributes.user_agent_original() => extract_user_agent(request)
+    }
 
     %{
-      Trace.http_method() => http_method(request.method),
-      Trace.http_url() => url,
-      Trace.http_target() => uri.path,
-      Trace.net_host_name() => uri.host,
-      Trace.http_scheme() => uri.scheme
+      HTTPAttributes.http_request_method() => parse_method(request.method),
+      ServerAttributes.server_address() => uri.host,
+      ServerAttributes.server_port() => extract_port(uri),
+      URLAttributes.url_full() => url
     }
-    |> maybe_append_req_content_length(request)
-    |> maybe_append_retry_count(request)
+    |> set_retry_count(request)
+    |> set_req_header_attrs(request)
+    |> then(fn attrs ->
+      opt_in
+      |> Map.take(config.opt_in_attrs)
+      |> Map.merge(attrs)
+    end)
   end
 
   defp sanitize_url(uri) do
     %{uri | userinfo: nil}
     |> URI.to_string()
   end
 
-  defp maybe_append_req_content_length(attrs, req) do
-    case Req.Request.get_header(req, "content-length") do
+  defp extract_port(%{port: port}) when is_integer(port), do: port
+
+  defp extract_port(%{scheme: scheme}) do
+    case scheme do
+      nil -> 80
+      "http" -> 80
+      "https" -> 8080
+      _ -> 80
+    end
+  end
+
+  defp extract_scheme(%{scheme: scheme}) do
+    case scheme do
+      nil -> :http
+      "http" -> :http
+      "https" -> :https
+      _ -> :http
+    end
+  end
+
+  defp extract_url_template(request) do
+    Req.Request.get_private(request, :path_params_template, "")
+  end
+
+  defp extract_user_agent(request) do
+    case Req.Request.get_header(request, "user-agent") do
       [] ->
-        attrs
+        ""
 
-      [length] ->
-        Map.put(attrs, Trace.http_request_content_length(), length)
+      [user_agent] ->
+        user_agent
     end
   end
 
-  defp maybe_append_resp_content_length(attrs, req) do
+  defp extract_response_body_size(req) do
     case Req.Response.get_header(req, "content-length") do
       [] ->
-        attrs
+        0
+
+      [length_str] when is_binary(length_str) ->
+        # Req sets this as a string but should be an integer
+        # https://www.rfc-editor.org/rfc/rfc9110.html#name-content-length
+        # https://opentelemetry.io/docs/specs/semconv/attributes-registry/http
+        String.to_integer(length_str)
+
+      [length] ->
```

Comment:
Similar to the above, and I imagine it would never happen in practice, but what if there were multiple headers returned? Makes me think maybe what we really want here is a `get_first_header` API where it returns the first one if there are one or more, or `nil` if it's not there, so make the call sites less likely to one day crash due to weird data.